### PR TITLE
CB-17761, CB-18252 - Implement database upgrade by replacing the RDS

### PR DESCRIPTION
**Pull request by pdarvasi**
This commit introduces the following upgrade logic for Azure:
 - deleting the database
 - deleting any private endpint if exists
 - updating db stack with new version
 - re-deploying the database
 - re-creating private endpoint if needed

The hence created database will be populated by restoring the previously dumped backup
data.

Was tested:
 - with and without Private Endpoints
 - on Data Lake and Data Hub
 - SSL enabled database
 - single RG setup
 - multiple RG setup

2E314958-88C1-4303-94B8-35F9CCB80826

### CODE REVIEW

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/connector/resource/AzureDatabaseResourceService.java [hunk 1]
```diff
@@ -238,4 +244,76 @@ public ExternalDatabaseStatus getDatabaseServerStatus(AuthenticatedContext ac, D
...
+                throw azureUtils.convertToCloudConnectorException(e, "Database stack upgrade");
+            } catch (Exception e) {
+                throw new CloudConnectorException(String.format("Error in upgrading database stack %s: %s", stackName, e.getMessage()), e);
+            } finally {
+                updateCloudResourcesInDeployment(persistenceNotifier, cloudContext, stackName, resourceGroupName, client);
+            }
+        } else {
+            String exceptionMessage = String.format("Azure database server cloud resource does not exist for stack %d!", stackId);
+            LOGGER.warn(exceptionMessage);
+            throw new CloudConnectorException(exceptionMessage);
```
Comment (lacikaaa) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/connector/resource/AzureDatabaseResourceService.java:284]: wouldn't this cause an issue if the pod is restarted after the db deleted but before it's created?

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/connector/resource/AzureDatabaseResourceService.java [hunk 2]
```diff
@@ -238,4 +244,76 @@ public ExternalDatabaseStatus getDatabaseServerStatus(AuthenticatedContext ac, D
...
+
+    private void updateCloudResourcesInDeployment(PersistenceNotifier persistenceNotifier, CloudContext cloudContext,
+            String deploymentName, String resourceGroupName, AzureClient client) {
+        Deployment deployment = client.getTemplateDeployment(resourceGroupName, deploymentName);
+        if (deployment != null) {
+            List<CloudResource> cloudResources = azureCloudResourceService.getDeploymentCloudResources(deployment);
+            LOGGER.debug("Deployment {} has been found with the following cloud resources: {}", deploymentName, cloudResources);
+            cloudResources.forEach(resource -> persistenceNotifier.notifyUpdate(resource, cloudContext));
+        } else {
+            LOGGER.debug("Deployment {} is not found, it should not happen", deploymentName);
```
Comment (lacikaaa) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/connector/resource/AzureDatabaseResourceService.java:296]: then maybe this should be at least a warn

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/view/AzureDatabaseServerView.java [hunk 3]
```diff
@@ -6,7 +6,7 @@
 public class AzureDatabaseServerView {
 
     @VisibleForTesting
-    static final String DB_VERSION = "dbVersion";
+    public static final String DB_VERSION = "dbVersion";
```
Comment (lacikaaa) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/view/AzureDatabaseServerView.java:9]: in this case you should remove `@VisibleForTesting`

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java [hunk 4]
```diff
@@ -51,15 +66,47 @@ public Selectable doAccept(HandlerEvent<UpgradeRdsUpgradeDatabaseServerRequest>
...
+            LOGGER.error(String.format("Database 'upgrade' poller stopped for stack: %s", stackDto.getName()), e);
+            result = upgradeFailedEvent(stackId, e);
+        } catch (PollerException e) {
+            LOGGER.error(String.format("Database 'upgrade' polling failed for stack: %s", stackDto.getName()), e);
+            result = upgradeFailedEvent(stackId, e);
+        } catch (Exception e) {
+            LOGGER.error(String.format("Exception during DB upgrade for stack: %s", stackDto.getName()), e);
+            result = upgradeFailedEvent(stackId, e);
         }
+        return result;
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java:92]: nit: you could just inline return

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java [hunk 5]
```diff
@@ -51,15 +66,47 @@ public Selectable doAccept(HandlerEvent<UpgradeRdsUpgradeDatabaseServerRequest>
...
+        StackEvent result;
+        StackDto stackDto = stackDtoService.getById(stackId);
         try {
-            StackDto stackDto = stackDtoService.getById(stackId);
             if (embeddedDatabaseService.isAttachedDiskForEmbeddedDatabaseCreated(stackDto)) {
-                LOGGER.debug("Starting embedded database upgrade...");
-                rdsUpgradeOrchestratorService.upgradeEmbeddedDatabase(stackDto);
+                result = upgradeEmbeddedDatabase(request, stackId, stackDto);
+            } else {
+                result = upgradeExternalDatabase(request, stackId, stackDto);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java:75]: this is not in sync with the condition. if the embedded db is on the root disk we would start an external db upgrade?

File: redbeams/src/main/java/com/sequenceiq/redbeams/repository/DBResourceRepository.java [hunk 6]
```diff
@@ -25,4 +26,10 @@ public interface DBResourceRepository extends JpaRepository<DBResource, Long> {
 
     @Query("SELECT count(r) > 0 FROM DBResource r WHERE r.dbStack.id = :stackId AND r.resourceName = :name AND r.resourceType = :type")
     boolean existsByStackAndNameAndType(@Param("stackId") Long stackId, @Param("name") String name, @Param("type") ResourceType type);
+
+    @Query("SELECT r FROM DBResource r WHERE r.dbStack.id = :stackId AND r.resourceStatus = :resourceStatus AND r.resourceType = :resourceType")
+    Optional<DBResource> findFirstByResourceStatusAndResourceTypeAndDbStack(
```
Comment (lacikaaa) [inline redbeams/src/main/java/com/sequenceiq/redbeams/repository/DBResourceRepository.java:31]: `First` seems a bit misleading as it doesn't look like from the query if there is an order by and limit

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java (hunk 5 above)
Comment (pdarvasi) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java:75]: Good point, I am not really sure, what should happen with embedded db's on the root disk, I assume there should be a validation not to let them in, @keyki @schfeca75 could you pls comment?

Comment (pdarvasi) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/rds/UpgradeRdsHandler.java]: Sorted out the 3 different conditions

**Pull request merged**